src/main/java/at/ac/tuwien/kr/alpha/common/atoms/ExternalAtom.java

```diff
@@ -91,29 +92,26 @@ public boolean isInternal() {
 
 	@Override
 	public List<VariableTerm> getBindingVariables() {
-		if (negated) {
-			// Negative literal has no binding variables.
-			return Collections.emptyList();
-		}
-		LinkedList<VariableTerm> bindingVariables = new LinkedList<>();
-		for (Term term : input) {
-			bindingVariables.addAll(term.getOccurringVariables());
-		}
-		bindingVariables.addAll(output);
-		return bindingVariables;
+		// If the external atom is negative, then all variables of input and output are non-binding
+		// and there are no binding variables (like for ordinary atoms).
+		// If the external atom is positive, then variables of output are binding.
+		return negated ? emptyList() : unmodifiableList(output);
```

Comment:
I just now realized that `output` is a `List<VariableTerm>` and I guess this will make problems if one wants to use an external atom whose output is partially grounded.
For example as in the rule
`outgoing13(X) :- &getLargeGraphEdges[](13,X).`
which would be an external atom without input and two outputs where we are currently just interested in outgoing edges of node `13`. Does this work with the current `ExternalAtom`? I assume it does not and the fix probably would require that `output` is a list of `Terms`. Here you then have to scan that list for variables and return only those in case the atom is not negated.

This has also implications for the code where the binding occurs, because not every output is a variable and also not all returned tuples match with the grounded part. Those two conditions must be taken into account there (this probably occurs in several subclasses).

Reply:
Turns out that I confused input and output in the grammar. I just pushed abe82e7 to interpret external atoms as `&p[I1, I2, ..., In](O1, O2, ..., On)` where `I` are input and `O` are output.

Indeed, I think that we should forbid to have anything else than *variables* in the output list (for now) for the sake of simplicity.

I would interpret `&p[a](b,X)` as a type of filter. E.g. we run the function behind `&p` with input `a` and then "unify" it with `b, X` effectively discarding all output that does not have `b` at position 1. Equivalently: `&p[a](Y,X), Y = b`.

We can add this in a back-compatible manner any time.

Reply:
Forbidding anything except variables is okay for now and I think that later we should have the code for filtering directly in the `NaiveGrounder`, so that all code dealing with variables and substitutions is in one place. This probably means that the current code dealing with substitutions at the external atoms will move, too.

I agree with your interpretation of bound variables as some kind of filter. Actually, this gives rise to a simple rewriting for allowing terms as output, just the same way as in your example.